src/rust/engine/fs/store/src/snapshot.rs

```diff
@@ -432,11 +340,128 @@ impl Snapshot {
       let child_directories = future03::try_join_all(child_directory_futures).await?;
 
       out_dir.set_directories(protobuf::RepeatedField::from_vec(child_directories));
+
+      Self::error_for_collisions(&store, &parent_path, &out_dir).await?;
       store.record_directory(&out_dir, true).await
     }
     .boxed()
   }
 
+  ///
+  /// Ensure merge is unique and fail with debugging info if not.
+  ///
+  async fn error_for_collisions(
+    store: &Store,
+    parent_path: &Path,
+    dir: &bazel_protos::remote_execution::Directory,
+  ) -> Result<(), String> {
+    // Attempt to cheaply check for collisions to bail out early if there aren't any.
+    let unique_count = dir
+      .get_files()
+      .iter()
+      .map(bazel_protos::remote_execution::FileNode::get_name)
+      .chain(
+        dir
+          .get_directories()
+          .iter()
+          .map(bazel_protos::remote_execution::DirectoryNode::get_name),
+      )
+      .collect::<HashSet<_>>()
+      .len();
+    if unique_count == (dir.get_files().len() + dir.get_directories().len()) {
+      return Ok(());
+    }
+
+    let file_details_by_name = dir
+      .get_files()
+      .iter()
+      .map(|file_node| async move {
+        let digest_proto = file_node.get_digest();
+        let header = format!(
+          "file digest={} size={}:\n\n",
+          digest_proto.hash, digest_proto.size_bytes
+        );
+
+        let digest_res: Result<hashing::Digest, String> = digest_proto.into();
+        let contents = store
+          .load_file_bytes_with(digest_res?, |mut bytes| {
+            const MAX_LENGTH: usize = 1024;
+            let content_length = bytes.len();
+            if content_length > MAX_LENGTH && !log_enabled!(log::Level::Debug) {
+              bytes = bytes.slice_to(MAX_LENGTH);
+              bytes.extend_from_slice(
+                format!(
+                  "\n... TRUNCATED contents from {}B to {}B \
+                  (Pass -ldebug to see full contents).",
+                  content_length, MAX_LENGTH
+                )
+                .as_bytes(),
+              );
+            }
+            String::from_utf8_lossy(bytes.to_vec().as_slice()).to_string()
```

Comment:
```suggestion
            String::from_utf8_lossy(&bytes).to_string()
```

Can use the `AsRef` impl of `Bytes` to avoid this copy

Reply:
Mm, good point. If CI continues flaking, I'll apply this. Otherwise, this is a fairly low traffic codepath, so I might let it slide.